Support runmode specific execution

buuhuu opened the pull request:
This changeset extends the UpdateInfo with a multi value string field for runmodes. If
they are specified the update actions of that particular UpdateInfo will only executed
when running with all of the specified runmodes. If nothing is specified the actions
will be executed on all runmodes.

Resolves #38

Commit:
Support runmode specific execution (#38)

This changeset extends the UpdateInfo with a multi value string field for runmodes. If
they are specified the update actions of that particular UpdateInfo will only executed
when running with all of the specified runmodes. If nothing is specified the actions
will be executed on all runmodes.

kwin commented on vault-upgrade-hook/src/main/java/biz/netcentric/vlt/upgrade/handler/SlingUtils.java:
```diff
@@ -23,12 +26,29 @@
+    public boolean hasRunModes(Set<String> requiredRunModes) {
+        Set<String> actualRunMode = getRunModes();
+        boolean hasAll = true;
+        for (Iterator<String> it = requiredRunModes.iterator(); it.hasNext() && hasAll; ) {
```
A simple https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#containsAll(java.util.Collection) would suffice, right?

kwin reviewed (changes requested):
Looks good in general but can you please adjust the documentation as well?

Commit:
Updated documentation and added samples (#38)

The documentation now contains an paragraph mentioning the possibility
to limit the execution on specific runmodes. Additionally a sample for running
on author only was added to the groovy sample package.

Commit:
Simplified runmode constraint (#38)

Using Collection#containsAll() in disjunction with Collection#isEmpty()

kwin commented on vault-upgrade-hook/src/main/java/biz/netcentric/vlt/upgrade/handler/SlingUtils.java:
```diff
@@ -35,12 +35,7 @@ public ResourceResolver getResourceResolver(Session session) {
-            hasAll = actualRunMode.contains(it.next());
-        }
-        return hasAll;
+        return requiredRunModes.isEmpty() || getRunModes().containsAll(requiredRunModes);
```
IMHO the first condition is not even necessary

buuhuu replied:
You are right, done and added a test case.

Commit:
Simplified runmode constraint (#38)

Using Collection#containsAll() only, also added a test case

phornig commented on vault-upgrade-hook/src/main/java/biz/netcentric/vlt/upgrade/handler/SlingUtils.java:
```diff
@@ -23,12 +26,24 @@
     }
 
+    public boolean hasRunModes(Set<String> requiredRunModes) {
+        return getRunModes().containsAll(requiredRunModes);
```
The readme says "To do so set the `runmodes` property to an array of applicable runmodes." Which to me means any runmode in the list is applicable. This implementation means all need to match. So either the documentation should be more clear or the implementation should be changed.
I'd suggest to change the implementation and require only one runmode to match as this will give more flexibility. Or get fancy and allow multiple runmodes in one value joined by a ".", e.g. "runmodes="[publish1.dev,publish2.test]".

```suggestion
        return !Collections.disjoint(getRunModes(), requiredRunModes);
```

buuhuu replied:
In fact I wanted keep it compatible to the way the OsgiInstaller/JcrInstaller looks up artefacts from the repository.

But you are right it will be a pain duplicating `UpgradeAction`s when they are meant to run only on an intersection of runmodes like we do with osgi configuration already. I will adapt the implementation accordingly.

kwin replied:
IMHO AND use cases are much more likely than OR use cases. I.e. more often I want a config to be only applied on instances having all run modes set rather than on instances having any of the given run modes set. Also this is more in line with the JCR Installer (https://sling.apache.org/documentation/bundles/jcr-installer-provider.html#runmode-support) and the File Installer (https://sling.apache.org/documentation/bundles/file-installer-provider.html#runmode-support) which also only support AND.

buuhuu replied:
Agreed. What do you think about the following to be more precise:

> To do so set the runmodes property to an array of required runmodes. If set the `UpgradeAction`s will only be executed when the environment runs in all of the given runmodes.

phornig replied:
Maybe more likely, but that case could be covered by using a composite runmode like "authordev". I don't see a workaround for the OR case.

What about the fancy variant, e.g. "runmodes="[publish1.dev,publish2.test]"?

buuhuu replied:
This would then be 
```
Iterator<String> it = requiredRunModes.iterator();
boolean hasMatchingSetOfRunModes = requiredRunModes.isEmpty();
while (!hasMatchingSetOfRunModes && it.hasNext()) {
    Collection<String> collectionOfRequiredRunModes = Arrays.asList(it.next().split("\\."));
    hasMatchingSetOfRunModes = getRunModes().containsAll(collectionOfRequiredRunModes);
}
return hasMatchingSetOfRunModes;
```

Commit:
Removed unnecessary modifiers (#38)

Declaring static methods final doesn't make sense here.

Commit:
Support disjunction and conjunctions of runmodes (#38)

Similar to the osgiinstaller we support conjunctive sets of runmodes but to
be more flexible we introduced also support for disjunctive sets.
